narwhal/primary/src/certificate_waiter.rs

```diff
@@ -3,226 +3,380 @@
+    name: PublicKey,
+    network: Arc<dyn PrimaryToPrimaryRpc>,
+    committee: Committee,
+    request: FetchCertificatesRequest,
+) -> FetchCertificatesResponse {
+    trace!("Start sending fetch certificates requests");
+    let request_interval = Duration::from_secs(5);
+    let mut peers: Vec<NetworkPublicKey> = committee
+        .others_primaries(&name)
+        .into_iter()
+        .map(|(_, _, network_key)| network_key)
+        .collect();
+    loop {
+        peers.shuffle(&mut ThreadRng::default());
+        let mut fut = FuturesUnordered::new();
+        for peer in peers.iter() {
+            fut.push(network.fetch_certificates(peer, request.clone()));
+            let mut interval = Box::pin(time::sleep(
+                request_interval + Duration::from_millis(ThreadRng::default().gen_range(0..1000)),
+            ));
+            tokio::select! {
+                res = fut.next() => match res {
+                    Some(Ok(resp)) => {
+                        return resp;
+                    }
+                    Some(Err(e)) => {
+                        debug!("Failed to fetch certificates: {e}");
+                    }
+                    None => {}
+                },
+                _ = &mut interval => {}
+            };
         }
     }
+}
 
-    fn update_metrics(&self, waiting_len: usize) {
-        self.metrics
-            .pending_elements_certificate_waiter
-            .with_label_values(&[&self.committee.epoch.to_string()])
-            .set(self.pending.len() as i64);
-        self.metrics
-            .waiting_elements_certificate_waiter
-            .with_label_values(&[&self.committee.epoch.to_string()])
-            .set(waiting_len as i64);
+#[instrument(level = "debug", skip_all)]
+async fn store_certificates_helper(
+    response: FetchCertificatesResponse,
+    committee: Committee,
+    certificate_store: &CertificateStore,
+    tx_certificates_loopback: Sender<Certificate>,
+    metrics: Arc<PrimaryMetrics>,
+) -> DagResult<()> {
+    trace!("Start sending fetched certificates to processing");
+    let mut waiters = Vec::new();
+    for certificate in response.certificates {
+        if waiters.len() == MAX_CERTIFICATES_TO_FETCH {
+            break;
+        }
+        waiters.push(certificate_store.notify_read(certificate.digest()));
+        // NOTE: currently this relies on core to verify and sanitize the certificate,
+        // and making sure all parents exist in storage. In future we may want to move the verification here.
+        if tx_certificates_loopback.send(certificate).await.is_err() {
+            warn!("Failed to send fetched certificate to processing");
+            return Err(DagError::ClosedChannel(
+                "tx_certificates_loopback".to_string(),
+            ));
+        }
     }
+
+    let waiters_len = waiters.len();
+    // Timeout should be enough to write the maximum number of certificates.
+    let mut timeout = Box::pin(time::sleep(Duration::from_secs(30)));
```

Comment:
That makes me a bit skeptical . Let's assume that the other node is sending us a certificate for which we are missing parents , we'll wait here for 30 seconds until we are able to fetch any other certificate. Even if the other node returns only 1 certificate of ex very high round for which we are missing parents they can significantly stall us here. I would like to see your & others thoughts here - and probably wrap this better in my head.

Reply:
Refactored to have core reply to certificate waiter with a message, so we do not timeout here anymore.